exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go

```diff
@@ -0,0 +1,220 @@
+	Cluster clusterv1.Cluster `json:"cluster"`
+}
+
+// BeforeClusterCreateResponse is the response of BeforeClusterCreate hook.
+// +kubebuilder:object:root=true
+type BeforeClusterCreateResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonResponse contains a Status and Message field common to all response types.
+	CommonResponse `json:",inline"`
+
+	// RetryAfterSeconds when set to a non-zero value signifies that the hook
+	// will be called again at a future time.
+	RetryAfterSeconds int `json:"retryAfterSeconds"`
+}
+
+// BeforeClusterCreate is the runtime hook that will be called right before a Cluster is created.
+func BeforeClusterCreate(*BeforeClusterCreateRequest, *BeforeClusterCreateResponse) {}
+
+// AfterControlPlaneInitializedRequest is the request of AfterControlPlaneInitialized hook.
+// +kubebuilder:object:root=true
+type AfterControlPlaneInitializedRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// The cluster object the lifecycle hook corresponds to.
+	Cluster clusterv1.Cluster `json:"cluster"`
+}
+
+// AfterControlPlaneInitializedResponse is the response of AfterControlPlaneInitialized hook.
+// +kubebuilder:object:root=true
+type AfterControlPlaneInitializedResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonResponse contains a Status and Message field common to all response types.
+	CommonResponse `json:",inline"`
+}
+
+// AfterControlPlaneInitialized is the runtime hook that will be called after the control plane is available for the first time.
+func AfterControlPlaneInitialized(*AfterControlPlaneInitializedRequest, *AfterControlPlaneInitializedResponse) {
+}
+
+// BeforeClusterUpgradeRequest is the request of the hook.
+// +kubebuilder:object:root=true
+type BeforeClusterUpgradeRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// The cluster object the lifecycle hook corresponds to.
+	Cluster clusterv1.Cluster `json:"cluster"`
+
+	// The current Kubernetes version of the cluster.
+	FromKubernetesVersion string `json:"fromKubernetesVersion"`
+	// The target Kubernetes version of upgrade.
+	ToKubernetesVersion string `json:"toKubernetesVersion"`
+}
+
+// BeforeClusterUpgradeResponse is the response of BeforeClusterUpgrade hook.
+// +kubebuilder:object:root=true
+type BeforeClusterUpgradeResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// CommonResponse contains a Status and Message field common to all response types.
+	CommonResponse `json:",inline"`
+
+	// RetryAfterSeconds when set to a non-zero value signifies that the hook
+	// needs to be retried at a future time.
+	RetryAfterSeconds int `json:"retryAfterSeconds"`
+}
+
+// BeforeClusterUpgrade is the runtime hook that will be called after a cluster.spec.version is upgraded and
+// before the updated version is propagated to the underlying objects.
+func BeforeClusterUpgrade(*BeforeClusterUpgradeRequest, *BeforeClusterUpgradeResponse) {}
+
+// AfterControlPlaneUpgradeRequest is the request of the hook.
```

Comment:
```suggestion
// AfterControlPlaneUpgradeRequest is the request of the AfterControlPlaneUpgrade hook.
```
nit (for consistency with other comments)